app/controllers/api/v1x0/actions_controller.rb

```diff
@@ -1,61 +1,31 @@
 module Api
   module V1x0
     class ActionsController < ApplicationController
-      include Api::V1x0::Mixins::IndexMixin
-      include Api::V1x0::Mixins::RBACMixin
-
-      ADMIN_OPERATIONS     = [Action::MEMO_OPERATION, Action::APPROVE_OPERATION, Action::DENY_OPERATION, Action::CANCEL_OPERATION].freeze
-      APPROVER_OPERATIONS  = [Action::MEMO_OPERATION, Action::APPROVE_OPERATION, Action::DENY_OPERATION].freeze
-      REQUESTER_OPERATIONS = [Action::CANCEL_OPERATION].freeze
-
-      before_action :read_access_check, :only => %i[show]
-      before_action :create_access_check, :validate_create_action, :only => %i[create]
+      include Mixins::IndexMixin
 
       def index
         req = Request.find(params.require(:request_id))
+        authorize(req, :policy_class => RequestPolicy)
 
-        collection(index_scope(req.actions))
+        collection(policy_scope(req.actions))
       end
 
       def show
         action = Action.find(params.require(:id))
+        authorize action
 
         json_response(action)
       end
 
       def create
-        action = ActionCreateService.new(params.require(:request_id)).create(params_for_create)
-
-        json_response(action, :created)
-      end
-
-      private
-
-      def validate_create_action
-        operation = params.require(:operation)
-        uuid = request.headers['x-rh-random-access-key']
-
-        valid_operation =
-          admin? && ADMIN_OPERATIONS.include?(operation) ||
-          approver? && APPROVER_OPERATIONS.include?(operation) ||
-          requester? && REQUESTER_OPERATIONS.include?(operation) ||
-          uuid.present? && RandomAccessKey.find_by(:access_key => uuid)
+        authorize Action
 
-        raise Exceptions::NotAuthorizedError, "Not authorized to create [#{operation}] action " unless valid_operation
-
-        resource_check('read', params[:request_id], Request) # NotAuthorizedError if current user cannot access the particular request
-      end
-
-      def rbac_scope(relation)
-        return relation if admin?
-
-        # Only approver can reach here
-        resource_check('read', params[:request_id], Request) # NotAuthorizedError if current user cannot access the particular request
-
-        action_ids = approver_id_list(relation.model.table_name)
-        Rails.logger.debug { "Approver scope for actions: #{action_ids}" }
+        # make sure have permission to read the request
+        req = Request.find(params.require(:request_id))
+        authorize(req, 'show?', :policy_class => RequestPolicy)
```

Comment:
@hsong-rh Why are we passing in the RequestPolicy all the time, can't we figure it out from the parent object. In Catalog we have Portfolio Item which is child of Portfolio. So in the Policy for PortfolioItem we just pick up the Portfolio.
https://github.com/RedHatInsights/catalog-api/blob/358797e376b284f9cdeab6038a1fcafcf9cb3287/app/policies/portfolio_item_policy.rb#L8

Reply:
Catalog did the same thing, from child to pick up parent, right? Catalog did the similar logic in child policy. Approval checks and may stop it in parent level.

Reply:
@hsong-rh We have Policies for each of the objects and do the appropriate thing from that Policy object.
e.g. https://github.com/RedHatInsights/catalog-api/tree/master/app/policies
The Policy object does the right thing it knows how to do the validation by looking at the appropriate parent object. This allows us to move the logic from the controller to the Policy objects and allows us to change the logic in the Policy object in the future.

Reply:
Ok, I will work to move the logic to its `policy` class.